web_src/js/features/contextpopup.js

```diff
@@ -1,45 +1,54 @@
-import {createApp} from 'vue';
 import ContextPopup from '../components/ContextPopup.vue';
+import {createVueRoot} from '../utils/vue.js';
 import {parseIssueHref} from '../utils.js';
 import {createTippy} from '../modules/tippy.js';
+import {GET} from '../modules/fetch.js';
 
-export function initContextPopups() {
-  const refIssues = document.querySelectorAll('.ref-issue');
-  attachRefIssueContextPopup(refIssues);
+const {appSubUrl} = window.config;
+const initAttr = 'data-contextpopup-init-done';
+
+async function init(e) {
+  const link = e.currentTarget;
+  if (link.hasAttribute(initAttr)) return;
+  link.setAttribute(initAttr, 'true');
+
+  const {owner, repo, index} = parseIssueHref(link.getAttribute('href'));
+  if (!owner) return;
+
+  const res = await GET(`${appSubUrl}/${owner}/${repo}/issues/${index}/info`); // backend: GetIssueInfo
+  if (!res.ok) return;
+
+  let issue, labelsHtml;
+  try {
+    ({issue, labelsHtml} = await res.json());
+  } catch {}
+  if (!issue) return;
+
+  const content = createVueRoot(ContextPopup, {issue, labelsHtml});
+  if (!content) return;
+
+  const tippy = createTippy(link, {
+    theme: 'default',
+    trigger: 'mouseenter focus',
+    content,
+    placement: 'top-start',
+    interactive: true,
+    role: 'tooltip',
+    interactiveBorder: 15,
+  });
+
+  // show immediately because this runs during mouseenter and focus
+  tippy.show();
 }
 
-export function attachRefIssueContextPopup(refIssues) {
-  for (const refIssue of refIssues) {
-    if (refIssue.classList.contains('ref-external-issue')) {
-      return;
-    }
-
-    const {owner, repo, index} = parseIssueHref(refIssue.getAttribute('href'));
-    if (!owner) return;
-
-    const el = document.createElement('div');
-    el.classList.add('tw-p-3');
-    refIssue.parentNode.insertBefore(el, refIssue.nextSibling);
-
-    const view = createApp(ContextPopup);
-
-    try {
-      view.mount(el);
-    } catch (err) {
-      console.error(err);
-      el.textContent = 'ContextPopup failed to load';
-    }
-
-    createTippy(refIssue, {
-      theme: 'default',
-      content: el,
-      placement: 'top-start',
-      interactive: true,
-      role: 'dialog',
-      interactiveBorder: 5,
-      onShow: () => {
-        el.firstChild.dispatchEvent(new CustomEvent('ce-load-context-popup', {detail: {owner, repo, index}}));
-      },
-    });
+export function attachRefIssueContextPopup(els) {
+  for (const el of els) {
+    el.addEventListener('mouseenter', init, {once: true});
+    el.addEventListener('focus', init, {once: true});
```

Comment:
It doesn't seem right to use "once".

For example, the first event call meets network error, then no issue info is shown, the the user could only refresh the whole page to make the popup work again.

Reply:
Just want to avoid introducing bugs like initializing two tippy instances etc. There is also `initAttr` because focus and mouseenter could happen both, so this `once` can be removed, but I still need a way to prevent double tippy, maybe just by checking existance of `el._tippy`.

Reply:
Fixed in https://github.com/go-gitea/gitea/pull/31181/commits/bc8dcbca3dd92a419dc94cf37f5389d5dcd7f5a2, only checking for `_tippy` which indicates the function ran to success already for the current link.

Reply:
And improved in https://github.com/go-gitea/gitea/pull/31181/commits/0ec0e43fc87f6c99445be80ce8f747f7fc8028e6, where it now also supports the case when the `href` on the link changes after tooltip creation, which currently can't happen but might in the future.